### PR TITLE
Install right version of cstruct rubygem before calling ohai

### PR DESCRIPTION
**Pull request by vuntz**
This ensures that the cstruct rubygem is installed, even after we
upgraded to a new version of ruby.

Note that this should actually help get rid of the hard-coding of this
gem in the provisioner barclamp.

### CODE REVIEW

**Comment by vuntz** [comment]
Note: this requires a rebuild of sleshammer.

**Comment by tboerger** [comment]
It is already part of sleshammer in OBS

**Comment by tboerger** [comment]
> We also tried a somewhat similar patch in the past. Which we needed to revert again:
> crowbar#102
> and
> crowbar#107

But you have not cleared the gem load paths

**Comment by rhafer** [comment]
> But you have not cleared the gem load paths

Yeah, but the issue really is the missing packaging stuff in sleshammer. Which is intentional because it save quite a bit of space. (See the recent discussion around that on our mailing list)

**Comment by tboerger** [comment]
> Yeah, but the issue really is the missing packaging stuff in sleshammer. Which is intentional because it save quite a bit of space. (See the recent discussion around that on our mailing list)

It is part of sleshammer. https://build.opensuse.org/package/view_file/systemsmanagement:crowbar:tex/sleshammer/sleshammer.SLE_11_SP3.config.xml?expand=1

**Comment by rhafer** [comment]
> It is part of sleshammer

I am still not quite sure what you mean by "it" :). Do you mean the rubygem-cstruct rpm? If yes, that's not enough. Chef will still try to call zypper/rpm to find out which version is installed. That will break as we remove rpm, zypper and the RPM database from the sleshammer image as part of the sleshammer built. At least in D:C:5 in IBS. And we do that on purpose.

If you mean something different, please clarify I might be misunderstanding you.

**Comment by vuntz** [comment]
I think @rhafer is right.

@rhafer are you fine if I protect all this with an `if` statement on the node state?

**Comment by rhafer** [comment]
> are you fine if I protect all this with an if statement on the node state?

That should work, yes. Please also add a comment to detail the reasons why we do it that way.

**Comment by tboerger** [comment]
But after gem clear paths the cstruct gem is still not required.

**Comment by vuntz** [comment]
@rhafer: updated.

@tboerger: cstruct is needed in the ohai plugin that we install, and that is then running at the end of the recipe.

**Comment by rhafer** [comment]
+1 This should work.

**Comment by vuntz** [comment]
Note: I added a new commit that was required to properly deal with the upgrade process.

**Comment by dirkmueller** [comment]
+1

File: chef/cookbooks/ohai/recipes/default.rb [hunk 1]
```diff
@@ -54,6 +54,30 @@
 end
 rd.run_action(:create)
 
+# we need to ensure that the cstruct gem is available (since we use it in our
+# plugin), except on sledgehammer (because it's already installed and we can't
+# install/check packages there)
+states = [ "ready", "readying", "recovering", "applying" ]
+if states.include?(node[:state])
+  # During the upgrade process (stoney -> tex, old ruby&rails -> nex
```
Comment (rhafer) [inline chef/cookbooks/ohai/recipes/default.rb]: typo: "nex"

Comment (vuntz) [inline chef/cookbooks/ohai/recipes/default.rb]: Fixed the typo.

**Pull request merged**